Fix typed arrays not working with concatenation operator `+`

mashumafi opened the pull request:
This fixes https://github.com/godotengine/godot/issues/72948 allowing concatenation of typed arrays and maintaining the type information. Includes a test case as well.

*Bugsquad edit:*
- Fixes #72948
- Fixes #72590

mashumafi commented:
This should also fix https://github.com/godotengine/godot/issues/72590

dalexeev commented:
How does it work with this?

```gdscript
var a: Array[int] = [1, 2, 3]
var b: Array[String] = ["one", "two", "three"]
print(a + b)
```

I think the sum type should be the narrowest type, including type a and type b, and not just the type of the first array? Or maybe it should cause an error?

mashumafi commented:
> How does it work with this?
> 
> ```gdscript
> var a: Array[int] = [1, 2, 3]
> var b: Array[String] = ["one", "two", "three"]
> print(a + b)
> ```
> 
> I think the sum type should be the narrowest type, including type a and type b, and not just the type of the first array? Or maybe it should cause an error?

This would likely result in an error currently. It wouldn't reach the concatenating portion of the code.
I think once `_add_arrays` is reached the types of both arrays were already determined to be identical. Perhaps adding some `CHECK` code to make that more clear but it does seem like the expectation. Open to ideas though. 

In the example the narrowest type would be `Varant`? It might be worth it to be able to write code like:
```gdscript
var node2ds : Array[Node2D] = [ ... ]
var node3ds : Array[Node3D] = [ ... ]
var nodes : Array[Node] = node2ds + node3ds
```

mashumafi commented:
Thinking about it more, perhaps adding the additional feature pushes this into 4.x territory. I think just focusing on this bug fix for now and open a separate issue for the narrowest type for a potential 4.x release.

mashumafi commented:
I took another look after understanding @dalexeev concern. Now the implementation is a bit more involved but covers more use-cases. Hopefully it's still in scope of 4.0. I feel pretty confident with because the tests pretty much cover all the functionality of mixing 2 types together. Feedback is appreciated!

akien-mga commented:
CC @vonagam

akien-mga commented:
> I feel pretty confident with because the tests pretty much cover all the functionality of mixing 2 types together

I don't see explicit tests for untyped arrays (though maybe some existing tests cover it)? That's my main concern with this change, fixing the obviously broken behavior for typed arrays makes sense but we shouldn't take any risk with untyped arrays at this late stage.

vonagam commented:
This one is problematic one.

I can approve of `Array[T] + Array[T] = Array[T]` but any other mixing should result in untyped array:
```gdscript
var nodes: Array[Node] = node2ds + node3ds # will work
var objects: Array[Object] = node2ds + node3ds # will fail
```

Trying to guess typed array element type will complicate and slow down things.

So in my opinion, there should no new `mix` method, addition to `_add_arrays` can be simple:
```cpp
if (array_a.is_typed() && array_a.is_same_typed(array_b)) {
  sum.set_typed(array_a.get_typed_builtin(), array_a.get_typed_class_name(), array_a.get_typed_script());
}
```

And ideally in analyzer [there](https://github.com/godotengine/godot/blob/6f64349bfe42b795a6c50e9c2093d38e7a992e7a/modules/gdscript/gdscript_analyzer.cpp#L2602) should be added handling of case with `Array[T] + Array[T]` so that the result is typed as `Array[T]` and not as an untyped one.

YuriSizov commented:
> This one is problematic one.

Okay, then let's give it a bit more time to develop and give you guys a bit more time to discuss it. We can cherry-pick the final solution into 4.0.x if it's safe enough.

mashumafi commented:
Adjusted the PR. Removed the mixing behavior, keeping it simple and simply adding types if both arrays match. Enhanced the analyzer code as well to return the proper type, works similar to the `_add_arrays` function, and now the editor shows the lines as "safe".

![image](https://user-images.githubusercontent.com/1059099/220180612-65c969df-0db2-4be5-81a2-bf54b9ac8558.png)

vonagam commented:
Superb. 

The other place worth updating is [`type_from_variant`](https://github.com/godotengine/godot/blob/9c960a8c2494eb826a557a7ffc96dd4547f9d31e/modules/gdscript/gdscript_analyzer.cpp#L4407) this will be used when two constant arrays combined together to get a datatype for typing. Right now it will be resolved to an untyped array.

One thing that you may stumble upon in tests - do not use empty constant typed arrays, there is an unresolved issue with how constants are stored.

vonagam commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -4740,6 +4740,12 @@ GDScriptParser::DataType GDScriptAnalyzer::get_operation_type(Variant::Operator
+	if (p_operation == Variant::OP_ADD && a_type == Variant::ARRAY && b_type == Variant::ARRAY) {
+		if (p_a.get_container_element_type() == p_a.get_container_element_type()) {
+			r_valid = true;
+			return p_a;
```
Small thing, `type_source` is an important part, let's say that `p_a` is a hard annotated type while `p_b` is a weak inferred one. Result should be a weak one.

So makes sense to move next lines of `bool hard_operation = p_a.is_hard_type() && p_b.is_hard_type();` and `GDScriptParser::DataType result;` before this if condition and have this inside condition:
```cpp
result = p_a;
result.type_source = hard_operation ? GDScriptParser::DataType::ANNOTATED_INFERRED : GDScriptParser::DataType::INFERRED;
```

mashumafi commented:
Updated with `hard_operation` and now concatenating `const` arrays results in green-lines

![image](https://user-images.githubusercontent.com/1059099/220200802-9f5bf7a7-8b79-4a2f-82eb-994b2cce4c24.png)

vonagam commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -4306,7 +4306,12 @@ GDScriptParser::DataType GDScriptAnalyzer::type_from_variant(const Variant &p_va
 	result.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT; // Constant has explicit type.
 
-	if (p_value.get_type() == Variant::OBJECT) {
+	if (p_value.get_type() == Variant::ARRAY && p_source->type == GDScriptParser::Node::BINARY_OPERATOR) {
```
Ideally `p_source` should not matter, it is for debugging purposes (warnings and errors).

To get an element type combo of `type_from_metatype` and relevant `make_builtin_meta_type`/`make_native_meta_type`/`make_script_meta_type` can be used.

mashumafi replied:
Thanks for the help, it looks much cleaner now!

vonagam replied:
`type_from_metatype` calls on args for `set_container_element_type` are missing.

mashumafi replied:
Oops, missed that, it's there now!

vonagam commented:
All good now.

dalexeev commented:
Looks good, except that the test does not match the code style (extra parentheses and spaces).

mashumafi committed:
Fix: Typed arrays aren't working with +

mashumafi commented:
> Looks good, except that the test does not match the code style (extra parentheses and spaces).

I took a look at some other tests and update mine to be closer in consistency :)

aw-square commented on core/variant/variant_op.h:
```diff
@@ -824,6 +824,11 @@ class OperatorEvaluatorAddArray {
 		int asize = array_a.size();
 		int bsize = array_b.size();
+
+		if (array_a.is_typed() && array_a.is_same_typed(array_b)) {
```
Do we need to check `array_b.is_typed()`?
Apologies for the drive by comment... :^) Testing a github action PR review bot and was using your PR as an example.

mashumafi replied:
Technically I don't even think `array_a.is_typed()` needs to be checked. `is_same_typed` might just be enough because if they are both untyped then `set_typed` pretty much acts like a no-op. I guess it can be a bit expensive to call `set_typed` just to set no type though so it helps optimize that use-case.

To answer your question, checking if `array_b.is_typed()` would be redundant. If `array_a.is_typed()` and `array_a.is_same_typed(array_b)` then it can be safe to assume `array_b.is_typed()`.

adamscott reviewed (approved):
As @vonagam and @dalexeev gave their thumbs up, I shall too.
Thanks @mashumafi! Sorry for the delay, by the way.

adamscott commented:
@akien-mga @YuriSizov We should maybe wait for @vnen input as this affects core. But I think everything is alright.

vnen reviewed (approved):
Looks good to me.

YuriSizov commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -4741,11 +4750,21 @@ GDScriptParser::DataType GDScriptAnalyzer::get_operation_type(Variant::Operator
 	bool hard_operation = p_a.is_hard_type() && p_b.is_hard_type();
+
+	if (p_operation == Variant::OP_ADD && a_type == Variant::ARRAY && b_type == Variant::ARRAY) {
+		if (p_a.get_container_element_type() == p_a.get_container_element_type()) {
```
Also it compares the value with itself...

YuriSizov replied:
See https://github.com/godotengine/godot/pull/78478
